pkg/tsdb/cloudwatch/response_parser.go

```diff
@@ -95,62 +103,84 @@ func parseGetMetricDataTimeSeries(metricDataResults map[string]*cloudwatch.Metri
-					Points: make([]tsdb.TimePoint, 0),
-				}
+				tags := map[string]string{multiValuedDimension: value}
 				for key, values := range query.Dimensions {
 					if key != multiValuedDimension && len(values) > 0 {
-						emptySeries.Tags[key] = values[0]
+						tags[key] = values[0]
 					}
 				}
 
-				emptySeries.Name = formatAlias(query, query.Stats, emptySeries.Tags, label)
-				result = append(result, &emptySeries)
+				emptyFrame := data.Frame{
+					Name: formatAlias(query, query.Stats, tags, label),
+					Meta: &data.FrameMeta{
+						Custom: map[string]interface{}{
+							"tags": tags,
+						},
+					},
+				}
+				frames = append(frames, &emptyFrame)
 			}
 		} else {
-			keys := make([]string, 0)
+			dims := make([]string, 0, len(query.Dimensions))
 			for k := range query.Dimensions {
-				keys = append(keys, k)
-			}
-			sort.Strings(keys)
-
-			series := tsdb.TimeSeries{
-				Tags:   make(map[string]string),
-				Points: make([]tsdb.TimePoint, 0),
+				dims = append(dims, k)
 			}
+			sort.Strings(dims)
 
-			for _, key := range keys {
-				values := query.Dimensions[key]
+			tags := map[string]string{}
+			for _, dim := range dims {
+				plog.Debug("Handling dimension", "dimension", dim)
+				values := query.Dimensions[dim]
 				if len(values) == 1 && values[0] != "*" {
-					series.Tags[key] = values[0]
+					plog.Debug("Got a tag value", "tag", dim, "value", values[0])
+					tags[dim] = values[0]
 				} else {
 					for _, value := range values {
 						if value == label || value == "*" {
-							series.Tags[key] = label
+							plog.Debug("Got a tag value", "tag", dim, "value", value, "label", label)
+							tags[dim] = label
 						} else if strings.Contains(label, value) {
-							series.Tags[key] = value
+							plog.Debug("Got a tag value", "tag", dim, "value", value, "label", label)
+							tags[dim] = value
 						}
 					}
 				}
 			}
 
-			series.Name = formatAlias(query, query.Stats, series.Tags, label)
-
+			timestamps := []float64{}
+			points := []*float64{}
 			for j, t := range metricDataResult.Timestamps {
 				if j > 0 {
 					expectedTimestamp := metricDataResult.Timestamps[j-1].Add(time.Duration(query.Period) * time.Second)
 					if expectedTimestamp.Before(*t) {
-						series.Points = append(series.Points, tsdb.NewTimePoint(null.FloatFromPtr(nil), float64(expectedTimestamp.Unix()*1000)))
+						timestamps = append(timestamps, float64(expectedTimestamp.Unix()*1000))
+						points = append(points, nil)
 					}
 				}
-				series.Points = append(series.Points, tsdb.NewTimePoint(null.FloatFrom(*metricDataResult.Values[j]),
-					float64(t.Unix())*1000))
+				val := metricDataResult.Values[j]
+				plog.Debug("Handling timestamp", "timestamp", t, "value", *val)
+				timestamps = append(timestamps, float64(t.Unix()*1000))
+				points = append(points, val)
+			}
+
+			fields := []*data.Field{
+				data.NewField("timestamp", nil, timestamps),
+				data.NewField("value", nil, points),
```

Comment:
I think maybe tags should be a data.Labels{}, and passed as the second argument to the value field. Part of the change to dataframes is that labels are actually part of our data model now.

Reply:
@kylebrandt Aha, thanks, I was totally unsure about the tags.